### ISSUE TITLE
Segmentation corpus incomplete?

### ISSUE THREAD

**Issue by jcboyd** [kermitt2/grobid#53]
Grobid experts,

When I process with the shipped Segmentation model, the performance is excellent--accuracy and F1 scores are close to perfect in all fields (this is reflected in the annotations themselves). However when I train a new model on the shipped Segmentation corpus, the evaluation is very poor. I can think of three possible reasons to explain this:

1) The training set shipped is only a small sample of the set used to train the shipped model
2) The number of l-bfgs iterations (6 by default) is too low
3) I have broken something :)

**Comment by kermitt2** [comment]
Hi !

The training set coming with the GitHub distribution is complete and should result in the same model as shipped. I just test it for the segmentation model with a fresh clone of the project and the new trained model is working fine. 

However having only 6 iterations with the Wapiti l-bfgs training is indeed not the normal behaviour... you should have 164 iterations to be precise. This happens because the stopping criteria is reached by error from the start of the training. The default window size for the devel stopping criterion is 5, so the training stops at the 6th iteration at the earliest. 

Two questions : 
- did you add new examples to the training data prior the training? you should have by default 120 examples and 17 labels for the segmentation model
- which OS/environment/CPU are you using?

See you!

**Comment by jcboyd** [comment]
Thank you for the swift response!

With the command to train,

java -Xmx1024m -jar grobid-trainer/target/grobid-trainer-0.3.3-SNAPSHOT.one-jar.jar 0 segmentation -gH grobid-home,

the output is the following:

* Summary
    nb train:    120
    nb labels:   17
    nb blocks:   605498
    nb features: 10293738
* Train the model with l-bfgs
  [   1] obj=120249,04  act=1330693  err=18,98%/100,00% time=14,48s/14,48s
  [   2] obj=101727,77  act=1189341  err=18,98%/100,00% time=8,27s/22,76s
  [   3] obj=85667,64   act=770651   err=18,98%/100,00% time=8,50s/31,25s
  [   4] obj=83485,54   act=815374   err=18,98%/100,00% time=8,54s/39,80s
  [   5] obj=80575,67   act=740944   err=18,98%/100,00% time=8,66s/48,46s
  [   6] obj=79421,06   act=779434   err=18,98%/100,00% time=8,79s/57,25s
* Save the model
* Done

Whether I run Grobid through the command line, with maven, or as an eclipse project I experience the same early stopping. I am running Ubuntu 14.04 with CPU Intel i5 x 4. Wapiti as a separate command line tool works as expected. If I adjust the stopping conditions in the Grobid source code there is no change. If I change the optimisation algorithm itself, I get the following error:

* Train the model with sgd-l1
    - Build the index
An irrecoverable stack overflow has occurred.
Please check if any of your loaded .so files has enabled executable stack (see man page execstack(8))
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f7dcd6ddc54, pid=10549, tid=140178648069888
#
# JRE version: OpenJDK Runtime Environment (7.0_75-b13) (build 1.7.0_75-b13)
# Java VM: OpenJDK 64-Bit Server VM (24.75-b04 mixed mode linux-amd64 compressed oops)
# Derivative: IcedTea 2.5.4
# Distribution: Ubuntu 14.04 LTS, package 7u75-2.5.4-1~trusty1
# Problematic frame:
# C  [libwapiti.so+0xbc54]  sgd_add+0x50
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#

For now I have changed the engine to CRF++ which runs perfectly.

Thanks again!

**Comment by kermitt2** [comment]
Thank you very much for the details! 

You can use indeed CRF++, but you will loose the advantages provided by Wapiti: faster training, more compact models and less memory used.

In my tests, the wapiti training is successful on recent Ubuntu and CentOS 64 bit environments. 

I suspect the following problem. The parameters for Wapiti are passed as a string during the JNI call (ok it's not the best practice ;). The epsilon parameter -e used as stopping criteria is by default 0.00001 and it is typically serialized 1.0E-5 or 1,0E-5 depending on your LOCALE. Possibly the LOCALE considered by Wapiti and the JVM are somehow different on your environment and Wapiti interprets the parameter 1,0E-5 as 1.0 or something similar. If epsilon is at 1.0, there will be only 6 iterations with l-bfgs training because the stopping criteria is immedialty reached, so it seems a possible reason.

However I could not reproduce your problem with the sgd training, it does not seem to be related to any parameter values.

A way to fix your problem with the l-bfgs training might be to enforce the way the parameter values are serialized, or to improve the JNI mapping for the parameters...

**Comment by kermitt2** [comment]
So I've tried to improve how the epsilon parameter is given to Wapiti in the latest commit 0824aa8cd630d49d850bbf6850fb18941bbac259

If you have a chance to update your Grobid and test again the training, I am interested in your feedback!

**Comment by jcboyd** [comment]
Thank you very much, this has resolved the early stopping issue!
